Algorithms/Search/BinarySearcher.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using Algorithms.Sorting;
+
+namespace Algorithms.Search
+{
+    public class BinarySearcher<T> : IEnumerator<T>
+    {
+        private readonly IList<T> _collection;
+        private readonly Comparer<T> _comparer;
+        private T _item;
+        private int _currentItemIndex;
+        private int _leftIndex;
+        private int _rightIndex;
+
+        /// <summary>
+        /// The value of the current item
+        /// </summary>
+        public T Current
+        {
+            get
+            {
+                return _collection[_currentItemIndex];
+            }
+        }
+
+        object IEnumerator.Current => Current;
+
+        /// <summary>
+        /// Class constructor
+        /// </summary>
+        /// <param name="collection">A list</param>
+        /// <param name="comparer">A comparer</param>
+        public BinarySearcher(IList<T> collection, Comparer<T> comparer)
+        {
+            if (collection == null)
+            {
+                throw new NullReferenceException("List is null");
+            }
+            _collection = collection;
+            _comparer = comparer;
+            HeapSorter.HeapSort(_collection);
+        }
+
+        public void SearchFor(T item)
+        {
+            if (item == null)
+            {
+                throw new NullReferenceException("Item to search for is not set");
+            }
+            Reset();
+            _item = item;
+        }
```

Comment:
Why do you have a separate function for setting the item to be searched? Can't the `BinarySearch()` method just assign the item internally and perform the search? What benefit do I get, as a user of this class, from having two functions and not just one?

Reply:
SearchFor() is nothing more than a setter for the item and I use it to test MoveNext() in isolation, without calling the BinarySearch(). I will try to find a different way to test these 2 methods separately.